tiledb/common/memory_tracker.h

```diff
@@ -151,6 +245,108 @@ class MemoryTracker {
 
   /** Memory usage by type. */
   std::unordered_map<MemoryType, uint64_t> memory_usage_by_type_;
+
+  /** The id of this MemoryTracker. */
+  uint64_t id_;
+
+  /** The type of this MemoryTracker. */
+  MemoryTrackerType type_;
+
+  /** The upstream memory resource. */
+  tdb::pmr::memory_resource* upstream_;
+
+  /** MemoryTrackerResource by MemoryType. */
+  std::unordered_map<MemoryType, std::shared_ptr<MemoryTrackerResource>>
+      resources_;
+
+  /** Memory counters by MemoryType. */
+  std::unordered_map<MemoryType, std::atomic<uint64_t>> counters_;
+
+  /** The total memory usage of this MemoryTracker. */
+  std::atomic<uint64_t> total_counter_;
+
+  /** Generate a unique id for this MemoryTracker. */
+  static uint64_t generate_id();
+};
+
+class MemoryTrackerManager {
+ public:
+  /** Constructor. */
+  MemoryTrackerManager() = default;
+
+  DISABLE_COPY_AND_COPY_ASSIGN(MemoryTrackerManager);
+  DISABLE_MOVE_AND_MOVE_ASSIGN(MemoryTrackerManager);
+
+  /**
+   * Create a new memory tracker.
+   *
+   * @return The created MemoryTracker.
+   */
+  shared_ptr<MemoryTracker> create_tracker();
+
+  /**
+   * Generate a JSON string representing the current state of tracked memory.
+   *
+   * @return A string containing the JSON representation of tracked memory.
+   */
+  std::string to_json();
+
+ private:
+  /** A mutext to protect our list of trackers. */
+  std::mutex mutex_;
+
+  /** A weak_ptr to the instances of MemoryTracker we create. */
+  std::vector<std::weak_ptr<MemoryTracker>> trackers_;
+};
+
+class MemoryTrackerReporter {
+ public:
+  /**
+   * Constructor.
+   *
+   * @param cfg The Config instance for the parent context.
+   * @param manager The MemoryTrackerManager instance on the context resources.
+   */
+  MemoryTrackerReporter(
+      const Config& cfg, shared_ptr<MemoryTrackerManager> manager)
+      : manager_(manager)
+      , filename_(cfg.get<std::string>("sm.memory.tracker.reporter.filename"))
```

Comment:
```suggestion
      , filename_(cfg.get<std::string>("sm.mem.tracker.reporter.filename"))
```

Reply:
I guess we are not documenting this, which might be fine for now. Does cfg.get return nullopt when this is not set?

Reply:
Correct, if the key is missing we get `std::nullopt` out which disables the reporter thread.